shell/platform/linux/fl_application.cc

```diff
@@ -0,0 +1,183 @@
+  gchar** dart_entrypoint_arguments;
+
+  // Default width of a Flutter window in pixels.
+  int window_width;
+
+  // Default height of a Flutter window in pixels.
+  int window_height;
+};
+
+#define FL_APPLICATION_GET_PRIVATE(app)                        \
+  ((FlApplicationPrivate*)fl_application_get_instance_private( \
+      FL_APPLICATION(app)))
+
+enum { kSignalRegisterPlugins, kSignalLastSignal };
+
+static guint fl_application_signals[kSignalLastSignal];
+
+G_DEFINE_TYPE_WITH_CODE(FlApplication,
+                        fl_application,
+                        GTK_TYPE_APPLICATION,
+                        G_ADD_PRIVATE(FlApplication))
+
+// Default implementation of FlApplication::register_plugins
+static void fl_application_register_plugins(FlApplication* self,
+                                            FlPluginRegistry* registry) {}
+
+// Implements GApplication::activate.
+static void fl_application_activate(GApplication* application) {
+  FlApplication* self = FL_APPLICATION(application);
+  FlApplicationPrivate* priv = FL_APPLICATION_GET_PRIVATE(self);
+
+  GtkApplicationWindow* window =
+      GTK_APPLICATION_WINDOW(gtk_application_window_new(GTK_APPLICATION(self)));
+
+  // Use a header bar when running in GNOME as this is the common style used
+  // by applications and is the setup most users will be using (e.g. Ubuntu
+  // desktop).
+  // If running on X and not using GNOME then just use a traditional title bar
+  // in case the window manager does more exotic layout, e.g. tiling.
+  // If running on Wayland assume the header bar will work (may need changing
+  // if future cases occur).
+  gboolean use_header_bar = TRUE;
+#ifdef GDK_WINDOWING_X11
+  GdkScreen* screen = gtk_window_get_screen(GTK_WINDOW(window));
+  if (GDK_IS_X11_SCREEN(screen)) {
+    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
+    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
+      use_header_bar = FALSE;
+    }
+  }
+#endif
+  if (use_header_bar) {
+    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
+    gtk_widget_show(GTK_WIDGET(header_bar));
+    gtk_header_bar_set_title(header_bar, priv->window_title);
+    gtk_header_bar_set_show_close_button(header_bar, TRUE);
+    gtk_window_set_titlebar(GTK_WINDOW(window), GTK_WIDGET(header_bar));
+  } else {
+    gtk_window_set_title(GTK_WINDOW(window), priv->window_title);
+  }
+
+  gtk_window_set_default_size(GTK_WINDOW(window), priv->window_width,
+                              priv->window_height);
+  gtk_widget_show(GTK_WIDGET(window));
+
+  g_autoptr(FlDartProject) project = fl_dart_project_new();
+  fl_dart_project_set_dart_entrypoint_arguments(
+      project, priv->dart_entrypoint_arguments);
+
+  FlView* view = fl_view_new(project);
+  gtk_widget_show(GTK_WIDGET(view));
+  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));
+
+  g_signal_emit(self, fl_application_signals[kSignalRegisterPlugins], 0,
+                FL_PLUGIN_REGISTRY(view));
```

Comment:
In a multi-window world we'll want plugin registration to happen once per engine (instead of once per view). Will we need to move this? Will that be hard to move later?

Reply:
See this design doc: https://docs.google.com/document/d/1WkJuzifgJXyceaSoCCu8qTETYmYBOYKoz7Qaxtsore4/edit#heading=h.sywe0ozapeg1

Reply:
Since it's done through this signal we can change this in FlApplication and there shouldn't be any change in the app required other than it will need to be multi-window aware if it wants to access the FlViews.

Reply:
Yes, to answer the original question it will be moved, but this wont be hard and should be transparent to the app.